node/cmd/ccq/utils.go

```diff
@@ -0,0 +1,231 @@
+	apiKey = strings.ToLower(apiKey)
+	permsForUser, exists := perms[strings.ToLower(apiKey)]
+	if !exists {
+		logger.Debug("invalid api key", zap.String("apiKey", apiKey))
+		return fmt.Errorf("invalid api key")
+	}
+
+	// TODO: Should we verify the signatures?
+
+	if len(qr.Signature) == 0 {
+		if !permsForUser.allowUnsigned || signerKey == nil {
+			logger.Debug("request not signed and unsigned requests not supported for apiKey",
+				zap.String("apiKey", apiKey),
+				zap.Bool("allowUnsigned", permsForUser.allowUnsigned),
+				zap.Bool("signerKeyConfigured", signerKey != nil),
+			)
+			return fmt.Errorf("request not signed")
+		}
+
+		// Sign the request using our key.
+		var err error
+		digest := query.QueryRequestDigest(env, qr.QueryRequest)
+		qr.Signature, err = ethCrypto.Sign(digest.Bytes(), signerKey)
+		if err != nil {
+			logger.Debug("failed to sign request", zap.String("apiKey", apiKey), zap.Error(err))
+			return fmt.Errorf("failed to sign request: %w", err)
+		}
+	}
+
+	var queryRequest query.QueryRequest
+	err := queryRequest.Unmarshal(qr.QueryRequest)
+	if err != nil {
+		logger.Debug("failed to unmarshal request", zap.String("apiKey", apiKey), zap.Error(err))
+		return fmt.Errorf("failed to unmarshal request: %w", err)
+	}
+
+	// Make sure the overall query request is sane.
+	if err := queryRequest.Validate(); err != nil {
+		logger.Debug("failed to validate request", zap.String("apiKey", apiKey), zap.Error(err))
+		return fmt.Errorf("failed to validate request: %w", err)
+	}
+
+	// Make sure they are allowed to make all of the calls that they are asking for.
+	for _, pcq := range queryRequest.PerChainQueries {
+		switch q := pcq.Query.(type) {
+		case *query.EthCallQueryRequest:
+			for _, callData := range q.CallData {
+				contractAddress, err := vaa.BytesToAddress(callData.To)
+				if err != nil {
+					logger.Debug("failed to parse contract address", zap.String("apiKey", apiKey), zap.String("contract", hex.EncodeToString(callData.To)), zap.Error(err))
+					return fmt.Errorf("failed to parse contract address: %w", err)
+				}
+				if len(callData.Data) < 4 {
+					logger.Debug("eth call data must be at least four bytes", zap.String("apiKey", apiKey), zap.String("data", hex.EncodeToString(callData.Data)))
+					return fmt.Errorf("eth call data must be at least four bytes")
+				}
+				call := hex.EncodeToString(callData.Data)
+				callKey := fmt.Sprintf("ethCall:%d:%s:%s", int(pcq.ChainId), contractAddress, call)
+				if _, exists := permsForUser.allowedCalls[callKey]; !exists {
+					logger.Debug("requested call not authorized", zap.String("apiKey", apiKey), zap.String("callKey", callKey))
+					return fmt.Errorf(`call "%s" not authorized`, callKey)
+				}
+			}
+		default:
+			logger.Debug("unsupported query type", zap.String("apiKey", apiKey), zap.Any("type", pcq.Query))
```

Comment:
I'm not sure that you need to have any logs for any of these errors if you're returning the errors. Just log in the caller.

That's an easy way to not print the apiKey secret over and over to the logs, even though it requires debug mode.

Reply:
I prefer these more detailed messages. Also, since they are debugs, they will rarely be logged.

Reply:
That works. I almost wish we had a trace log level that was more just for devs for stuff like this.